src/scenario/scenario.c

```diff
@@ -7,36 +7,124 @@
 {
     return scenario.is_saved;
 }
 
+static void state_offsets_init()
+{
+    state_offsets.size = 0; // First value in the state is the size of the state in bytes
+    state_offsets.start_info = state_offsets.size + sizeof(int32_t);
+    state_offsets.old_filler_1 = state_offsets.start_info + 14;
+    state_offsets.invasions_part1 = state_offsets.old_filler_1 + 160;
+    state_offsets.funds_enemy = state_offsets.invasions_part1 + (MAX_INVASIONS * 10); // 378
+    state_offsets.map_size = state_offsets.funds_enemy + 14;
+    state_offsets.briefing = state_offsets.map_size + 16; 
+    state_offsets.old_filler_2 = state_offsets.briefing + MAX_BRIEF_DESCRIPTION + MAX_BRIEFING; // 994
+    state_offsets.image = state_offsets.old_filler_2 + 20;
+    state_offsets.herds = state_offsets.image + 6;
+    state_offsets.demand_changes_part1 = state_offsets.herds + (MAX_HERD_POINTS * 4);
+    state_offsets.price_changes_part1 = state_offsets.demand_changes_part1 + (MAX_DEMAND_CHANGES * 9);
+    state_offsets.glademp = state_offsets.price_changes_part1 + (MAX_PRICE_CHANGES * 6); // 1336
+    state_offsets.random_events = state_offsets.glademp + 16;
+    state_offsets.fishing = state_offsets.random_events + 36;
+    state_offsets.old_filler_3 = state_offsets.fishing + (MAX_FISH_POINTS * 4);
+    state_offsets.invasions_part2 = state_offsets.old_filler_3 + 20;
+    state_offsets.old_filler_4 = state_offsets.invasions_part2 + (MAX_INVASIONS * 1); // 1460
+    state_offsets.rome_wheat = state_offsets.old_filler_4 + 80;
+    state_offsets.allowed_buildings = state_offsets.rome_wheat + 4;
+    state_offsets.win_criteria = state_offsets.allowed_buildings + (MAX_ALLOWED_BUILDINGS * 2); // 1644
+    state_offsets.map_points = state_offsets.win_criteria + 52;
+    state_offsets.invasions_part3 = state_offsets.map_points + 12;
+    state_offsets.misc = state_offsets.invasions_part3 + (MAX_INVASION_POINTS * 4); // 1740
+    state_offsets.custom_name = state_offsets.misc + 51; // 1791
+    state_offsets.end = state_offsets.custom_name + 51; // 1842. Not sure how custon name length was determined.
+}
+
+int scenario_get_state_buffer_size_by_savegame_version(int savegame_version)
+{
+    if (savegame_version <= SAVE_GAME_LAST_UNVERSIONED_SCENARIOS) {
+        return 1720;
+    } else if (savegame_version <= SAVE_GAME_LAST_NO_EXTENDED_REQUESTS) {
+        return 1838;
+    } else {
+        state_offsets_init();
+        return state_offsets.end;
+    }
+}
+
+int scenario_get_state_buffer_size_by_scenario_version(int scenario_version)
+{
+    if (scenario_version <= SCENARIO_LAST_UNVERSIONED) {
+        return 1720;
+    } else if(scenario_version <= SCENARIO_LAST_NO_WAGE_LIMITS) {
+        return 1830;
+    } else if (scenario_version <= SCENARIO_LAST_NO_EXTENDED_REQUESTS) {
+        return 1838;
+    } else {
+        state_offsets_init();
+        return state_offsets.end;
+    }
+}
+
 void scenario_save_state(buffer *buf)
 {
+    state_offsets_init();
```

Comment:
Since you call `get_state_buffer_size` below, I don't think you need this here.

Reply:
You are correct. Removed.